### ISSUE TITLE
False positive in `unconditional-recursion` in presence of func literals

### ISSUE THREAD

**Issue by chavacava** [mgechev/revive#902]
**Describe the bug**
The rule `unconditional-recursion` produces false positives when it finds calls to `foo` in function literals defined inside `foo`.
For example:

```go
func (c *connector) SpanStats(ctx context.Context, req *roachpb.SpanStatsRequest) (*roachpb.SpanStatsResponse, error) {
        // ... 
        err := c.withClient(ctx, func(ctx context.Context, c *client) error {
                stats, err := c.SpanStats(ctx, req)  // <- this call is wrongly flagged as unconditional recusive
               // ...
                return nil
        })
        return response, err
}
```

**To Reproduce**
1. I updated revive `go install github.com/mgechev/revive@latest`
2. Include the following code in `testdata/unconditional-recursion.go`
```go
func falsePositiveFuncLiteral() {
        _ = foo(func() {
                falsePositiveFuncLiteral()
        })
}
```
3. run `make test` (test will fail for `unconditional-recursion` rule)

**Expected behavior**
The new test pass.

**Comment by mfederowicz** [comment]
@chavacava to be clear what is to fix content of unconditional-recursion.go, or writing new test which asserts other testcases than actual one, or maybe both?

**Comment by chavacava** [comment]
The goal is to add the new test *and* fix the rule implementation to make tests pass.

**Comment by mfederowicz** [comment]
@chavacava we waiting for something more in this issue, other tests or something?

**Comment by chavacava** [comment]
Closed by #909
